### PR TITLE
Implement Code Generation for Constants

### PR DESCRIPTION
**Pull request by Danaozhong**
- This commit will implement the generation of constants, by
  doing the following:
- The value expression of consts are now evaluated.
- String literal expressions are now supported.
- Constants are now evaluated.
- Constant names are capitalized.
- Constant types are actually generated in Go.

### CODE REVIEW

File: internal/ast/expression.go [hunk 1]
```diff
@@ -578,6 +578,30 @@ func (expr *Expression) evaluateBitwiseExpression() error {
...
+	// remove the quotes
+	if len(expressionString) < 2 {
+		return errors.New("string literal should contain quotes")
+	}
+	expr.ResultType = ExpressionTypeString
+	expr.ResultStringValue = expressionString[1 : len(expressionString)-1]
+	return nil
+}
+
+func (expr *Expression) evaluateBinaryLiteral() error {
```
[the remark this answers is not in the document]
Comment (Danaozhong) [reply internal/ast/expression.go:592]: Added.

File: internal/ast/expression_test.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+				if !cmp.Equal(test.expectedError.Error(), err.Error()) {
+					t.Errorf("result error doesn't match (expected: %v, actual: %v", test.expectedError, err)
+				}
+			} else {
+				require.Equal(t, ast.ExpressionTypeString, test.testExpression.ResultType)
+				require.Equal(t, ast.EvaluationStateComplete, test.testExpression.EvaluationState)
+				if diff := cmp.Diff(test.expectedValue, test.testExpression.ResultStringValue); diff != "" {
+					t.Errorf("do not match: (-expected +got)\n%s", diff)
+				}
+			}
```
Comment (aignas) [inline internal/ast/expression_test.go:56]: nit:
```suggestion
				assert.EqualError(t, err.Error(), test.expectedError.Error())
				return
			}
			require.Equal(t, ast.ExpressionTypeString, test.testExpression.ResultType)
			require.Equal(t, ast.EvaluationStateComplete, test.testExpression.EvaluationState)
			assert.Equal(t, test.expectedValue, test.testExpression.ResultStringValue)
```

The error equality could use the above function and we can make an early return. We could also cleanup the assertions.

File: internal/ast/expression.go [hunk 3]
```diff
@@ -722,8 +741,9 @@ func (expr *Expression) Evaluate(scope *Package) error {
 		err = errors.New("unknown expression type")
 	}
 
-	if err == nil {
-		expr.EvaluationState = EvaluationStateComplete
+	if err != nil {
+		return fmt.Errorf("evaluate %+v: %w", expr, err)
```
Comment (aignas) [inline internal/ast/expression.go:745]: What about just using the string representation of the expression here?
```suggestion
		return fmt.Errorf("evaluate %q: %w", expr.Text, err)
```

**Review by aignas — APPROVED** [review]
Thanks for adding the tests

**Pull request merged**